File: 2015/day23.py
from typing import Callable
from os.path import isfile, join as path_join

file_name = path_join('input', 'day23.txt')


def to_list(mf: Callable = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]


def to_gen(mf: Callable = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)


if not isfile(file_name):
	from aoc import get_input_file

	get_input_file()

from enum import IntEnum


class OpType(IntEnum):
	hlf = 0,
	tpl = 1,
	inc = 2,
	jmp = 3,
	jie = 4,
	jio = 5


class VM:
	reg_index = {"a": 0, "b": 1}

	def __init__(self, instructions):
		self.instructions = [tuple()] * len(instructions)
		self.reg = [0, 0]
		self.pc = 0

		for i, instr in enumerate(instructions):
			parts = instr.split()
			if parts[0] == "hlf":
				op = (OpType.hlf, VM.reg_index[parts[1]])
			elif parts[0] == "tpl":
				op = (OpType.tpl, VM.reg_index[parts[1]])
			elif parts[0] == "inc":
				op = (OpType.inc, VM.reg_index[parts[1]])
			elif parts[0] == "jmp":
				op = (OpType.jmp, int(parts[1]))
			elif parts[0] == "jie":
				op = (OpType.jie, VM.reg_index[parts[1][:-1]], int(parts[2]))
			elif parts[0] == "jio":
				op = (OpType.jio, VM.reg_index[parts[1][:-1]], int(parts[2]))

			self.instructions[i] = op

	def run(self):
		while self.pc < len(self.instructions):
			instr = self.instructions[self.pc]

			op = instr[0]
			operands = instr[1:]

			if op == OpType.hlf:
				self.reg[operands[0]] //= 2
				self.pc += 1
			elif op == OpType.tpl:
				self.reg[operands[0]] *= 3
				self.pc += 1
			elif op == OpType.inc:
				self.reg[operands[0]] += 1
				self.pc += 1
			elif op == OpType.jmp:
				self.pc += operands[0]
			elif op == OpType.jie:
				if self.reg[operands[0]] % 2 == 0:
					self.pc += operands[1]
				else:
					self.pc += 1
			elif op == OpType.jio:
				if self.reg[operands[0]] == 1:
					self.pc += operands[1]
				else:
					self.pc += 1


instr = open(file_name).read().strip().split("\n")


def part1():
	vm = VM(instr)
	vm.run()
	return vm.reg[VM.reg_index["b"]]


def part2():
	vm = VM(instr)
	vm.reg[VM.reg_index["a"]] = 1
	vm.run()
	return vm.reg[VM.reg_index["b"]]


print(part1())
print(part2())

File: cpp/2015/day23.cpp
#include "../aocHelper.h"

class Day23 : public BaseDay
{
public:
	Day23() : BaseDay("23") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		enum class opcode : int
		{
			nop,
			hlf,
			tpl,
			inc,
			jmp,
			jie,
			jio
		};

		struct operands
		{
			int op1 = 0;
			int op2 = 0;
		};

		struct instruction
		{
			opcode type = opcode::nop;
			operands data;
		};

		array<long long, 2> registers{};

		vector<instruction> instructions;

		// parse input
		while (*input != '\0')
		{
			instruction instr;

			if (*input == 'h') // hlf
			{
				input += 4; // skip 'hlf '
				instr.type = opcode::hlf;
				instr.data.op1 = *input - 'a';
				input++;
			}
			else if (*input == 't') // tpl
			{
				input += 4; // skip 'tpl '
				instr.type = opcode::tpl;
				instr.data.op1 = *input - 'a';
				input++;
			}
			else if (*input == 'i') // inc
			{
				input += 4; // skip 'inc '
				instr.type = opcode::inc;
				instr.data.op1 = *input - 'a';
				input++;
			}
			else if (*input == 'j')
			{
				input++;
				if (*input == 'm') // jmp
				{
					input += 3; // skip 'mp '
					instr.type = opcode::jmp;
					bool neg = (*input == '-');
					input++;
					instr.data.op1 = numericParse<int>(input);
					if (neg)
					{
						instr.data.op1 *= -1;
					}
				}
				else if (*input == 'i')
				{
					input++;
					if (*input == 'e') // jie
					{
						input += 2; // skip 'e '
						instr.type = opcode::jie;
						instr.data.op1 = *input - 'a';
						input += 3;

						bool neg = (*input == '-');
						input++;
						instr.data.op2 = numericParse<int>(input);
						if (neg)
						{
							instr.data.op2 *= -1;
						}
					}
					else if (*input == 'o') // jio
					{
						input += 2; // skip 'o '
						instr.type = opcode::jio;
						instr.data.op1 = *input - 'a';
						input += 3;

						bool neg = (*input == '-');
						input++;
						instr.data.op2 = numericParse<int>(input);
						if (neg)
						{
							instr.data.op2 *= -1;
						}
					}
				}
			}

			instructions.push_back(instr);

			input++; // skip \n
		}

		auto run_vm = [](array<long long, 2>& registers, vector<instruction>& instructions)
		{
			int pc = 0;

			while (true)
			{
				if (pc < 0 || pc >= instructions.size())
				{
					break;
				}

				instruction& instr = instructions[pc];

				switch (instr.type)
				{
					case opcode::nop:
					{
						cout << "nop" << endl;
						pc++;
						break;
					}
					case opcode::hlf:
					{
						registers[instr.data.op1] /= 2;
						pc++;
						break;
					}
					case opcode::tpl:
					{
						registers[instr.data.op1] *= 3;
						pc++;
						break;
					}
					case opcode::inc:
					{
						registers[instr.data.op1]++;
						pc++;
						break;
					}
					case opcode::jmp:
					{
						pc += instr.data.op1;
						break;
					}
					case opcode::jie:
					{
						if (registers[instr.data.op1] % 2 == 0)
						{
							pc += instr.data.op2;
						}
						else
						{
							pc++;
						}
						break;
					}
					case opcode::jio:
					{
						if (registers[instr.data.op1] == 1)
						{
							pc += instr.data.op2;
						}
						else
						{
							pc++;
						}
						break;
					}
				}
			}
		};

		// part 1
		run_vm(registers, instructions);

		part1 = registers[1];

		// part 2
		registers = { 1, 0 };
		run_vm(registers, instructions);

		part2 = registers[1];

		return { part1, part2 };
	}
};
